### PR TITLE
feat: create new grid area from integration event

### PR DESCRIPTION
**Pull request by MadsDue**
## Description
Create new grid area when market participant raise 'GridAreaOwnershipAssigned' integration event.

- [x] new integration event command mapper for event
- [x] raise CreateGridAreaCommand
- [x] handle command and persist the new gridarea

Next PR: 
- [Start using the new grid area table to lookup gridareas and clean up old ways](https://github.com/Energinet-DataHub/opengeh-edi/pull/591)

## References
https://app.zenhub.com/workspaces/mosaic-60a6105157304f00119be86e/issues/gh/energinet-datahub/team-phoenix/87

### CODE REVIEW

File: source/Application/GridAreas/CreateGridAreaCommand.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+using Energinet.DataHub.EDI.Application.Configuration.Commands.Commands;
+using Energinet.DataHub.EDI.Domain.Actors;
+using NodaTime;
+
+namespace Energinet.DataHub.EDI.Application.GridAreas;
+
+public class CreateGridAreaCommand : InternalCommand
+{
+    [JsonConstructor]
+    public CreateGridAreaCommand(string gridAreaCode, Instant validFrom, ActorNumber actorNumber)
```
Comment (lasrinnil) [inline source/Application/GridAreas/CreateGridAreaCommand.cs:25]:
```suggestion
    public CreateGridAreaCommand(string gridAreaCode, Instant validFrom, ActorNumber gridAreaOwnerActorNumber)
```

File: source/Infrastructure/Configuration/CompositionRoot.cs [hunk 2]
```diff
@@ -254,6 +257,12 @@ private void AddActorServices()
             _services.AddTransient<IActorRegistry, ActorRegistry>();
         }
 
+        private void AddGridAreaServices()
+        {
+            _services.AddTransient<IRequestHandler<CreateGridAreaCommand, Unit>, CreateGridAreaHandler>();
```
Comment (lasrinnil) [inline source/Infrastructure/Configuration/CompositionRoot.cs:262]: Think Scoped lifetime would suffice

File: source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/GridAreaOwnershipAssignedMapper.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+using Energinet.DataHub.EDI.Application.Actors;
+using Energinet.DataHub.EDI.Application.Configuration.Commands.Commands;
+using Energinet.DataHub.EDI.Application.GridAreas;
+using Energinet.DataHub.EDI.Domain.Actors;
+using Energinet.DataHub.MarketParticipant.Infrastructure.Model.Contracts;
+using NodaTime.Serialization.Protobuf;
+
+namespace Energinet.DataHub.EDI.Infrastructure.Configuration.IntegrationEvents.IntegrationEventMappers;
+
+public class GridAreaOwnershipAssignedMapper : IIntegrationEventMapper
```
Comment (lasrinnil) [inline source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/GridAreaOwnershipAssignedMapper.cs:26]:
```suggestion
public class GridAreaOwnershipAssignedEventMapper : IIntegrationEventMapper
```

Comment (lasrinnil) [inline source/Infrastructure/Configuration/IntegrationEvents/IntegrationEventMappers/GridAreaOwnershipAssignedMapper.cs:26]: For me, something is off here when we receive an event detailing that ownership of a grid area is assigned, and then we create a whole new grid area. Maybe it is just me :)

File: source/IntegrationTests/Infrastructure/IntegrationEvents/WhenGridAreaOwnershipAssignedTests.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        var gridAreaOwnershipAssignedEvent = GridAreaOwnershipAssignedEventBuilder.Build();
+
+        await HavingReceivedAndHandledIntegrationEventAsync(GridAreaOwnershipAssigned.EventName, gridAreaOwnershipAssignedEvent);
+
+        var gridAreas = await GetGridAreas(gridAreaOwnershipAssignedEvent.GridAreaCode, gridAreaOwnershipAssignedEvent.ActorNumber);
+        Assert.Single(gridAreas);
+    }
+
+    [Fact]
+    public async Task New_grid_area_is_received_with_existing_grid_area_code_does_not_store_second_grid_area()
```
Comment (lasrinnil) [inline source/IntegrationTests/Infrastructure/IntegrationEvents/WhenGridAreaOwnershipAssignedTests.cs:53]:
```suggestion
    public async Task New_grid_area_event_is_received_with_existing_grid_area_code_does_not_store_second_grid_area()
```

File: source/Domain/GridAreas/GridArea.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    {
+    }
+
+    public string GridAreaCode { get; }
+    public Instant ValidFrom { get; private set; }
+    public ActorNumber GridAreaOwnerActorNumber { get; private set; }
+
+    public void UpdateOwnership(Instant validFrom, ActorNumber actorNumber)
+    {
+        if(ValidFrom > validFrom) return;
```
Comment (RasmusGraabaek) [inline source/Domain/GridAreas/GridArea.cs:43]: Say actor1 has ownership of the grid area and it is supposeed to be assigned to actor2 witnin 2 years.
Then it is not possible for any other actor than actor1 to have ownership of the grid area within thise 2 years?

Comment (MadsDue) [reply source/Domain/GridAreas/GridArea.cs:43]: That is not a case :)

File: source/IntegrationTests/Infrastructure/IntegrationEvents/WhenGridAreaOwnershipAssignedTests.cs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        var gridAreaOwnershipAssignedEvent = _gridAreaOwnershipAssignedEventBuilder.Build();
+
+        await HavingReceivedAndHandledIntegrationEventAsync(GridAreaOwnershipAssigned.EventName, gridAreaOwnershipAssignedEvent);
+
+        var gridAreas = await GetGridAreas();
+        Assert.Single(gridAreas);
+    }
+
+    [Fact]
+    public async Task Multiple_grid_area_ownership_assigned_event_is_received_with_same_owner_is_stores()
```
Comment (RasmusGraabaek) [inline source/IntegrationTests/Infrastructure/IntegrationEvents/WhenGridAreaOwnershipAssignedTests.cs:55]:
```suggestion
    public async Task Multiple_grid_area_ownership_assigned_event_is_received_with_same_owner_is_stored()
```

**Review by RasmusGraabaek — APPROVED** [review]

**Pull request merged**